doc/tutorial/mpi.rst

```diff
@@ -0,0 +1,115 @@
+.. _tutorialmpi:
+
+MPI
+==============
+
+In this example, the ring network created in an :ref:`earlier tutorial <tutorialnetworkring>` will be used to run in a distributed context using MPI. Only the difference with that tutorial will be described.
+
+.. Note::
+
+   **Concepts covered in this example:**
+
+   1. Building a basic MPI aware :py:class:`arbor.context` to run a network.
+      This requires that you have built Arbor with MPI support enabled.
+   2. Running the simulation and extract the results.
+
+The recipe
+**********
+
+Step **(11)** is changed to generate a network with a thousand cells.
+
+.. code-block:: python
+
+   # (11) Instantiate recipe
+   ncells = 500
+   recipe = ring_recipe(ncells)
+
+The hardware context
+********************
+
+Step **(12)** uses the Arbor-built-in :py:class:`MPI communicator <arbor.mpi_comm>`, which identical to the
+``COMM_WORLD`` communicator you'll know if you are familiar with MPI. The :py:class:`arbor.context` takes it as an
+argument to the ``mpi`` parameters. Note that you can also pass in communicators created with ``mpi4py``.
+We print both the communicator and context to observe how Arbor configures their defaults.
+
+.. code-block:: python
+
+   # (12) Create an MPI communicator, and use it to create a hardware context
+   comm = arbor.mpi_comm()
+   print(comm)
+   context = arbor.context(mpi=comm)
+   print(context)
+
+The execution
+*************
+
+Step **(16)** runs the simulation. Since we have more cells this time, which are connected in series, it will take some time for the action potential to propagate. In the :ref:`ring network <tutorialnetworkring>` we could see it takes about 5 ms for the signal to propgate through one cell, so let's set the runtime to ``5*ncells``.
+
+.. code-block:: python
+
+   # (16) Run simulation
+   sim.run(ncells*5)
+   print('Simulation finished')
+
+An important change in the execution is how the script is run. Whereas normally you run the Python script by passing it as an argument to the ``python`` command, you need to use ``srun`` or ``mpirun`` (depending on your MPI distribution) to execute a number of jobs in parallel. You can still execute the script using ``python``, but then MPI will not execute on more than one node.
+
+From the commandline, we can run the script using ``mpirun`` or ``srun`` and specify the number of ranks (``NRANKS``) or nodes. Arbor will spread the cells evenly over the ranks, so with ``NRANKS`` set to 5, we'd be spreading the 500 cells over 5 nodes, calculating 100 cells each.
+
+.. code-block::
+
+   srun -n NRANKS python network_ring_mpi.py
+   mpirun -n NRANKS python network_ring_mpi.py
+
+The results
+***********
+
+Before we execute the simulation, we have to understand how Arbor distributes the computational load over the ranks
+and executes the rest of the script. You can visualize this as a fork from the moment :py:func:`arbor.simulation.run`
```

Comment:
I don't know about the fork once `run` is reached. The whole script is executed in parallel across all the machines. Even the partitioning happens in parallel. So I think this is misleading.

Reply:
The flow of the script forks at this point. From here I can expect execution to be different between nodes, and handles to be useless etc. How would you describe this, if not fork?

Reply:
I think it's misleading. There's no fork, every rank is running an independent process. When we launch the simulation with `mpirun` on let's say 10 ranks, the script is being run on each of these 10 ranks. When we reach `partition_load_balance` that's when the work is being distributed according to the `rank_id`. So rank 0 takes the 0'th chunk of cell, rank 1 takes the 1st chunk and so on. By the time we reach `sim.run` each rank knows which cells it is supposed to be simulating. 
The problem arises when when we get to assigning `handles`. We are generating handles for all the cells on all the cells on each rank. But not all of them are available, that's why we get empty handles later on. I'll address this issue and how to solve it on slack.